test.cpp

```diff
@@ -253,6 +253,114 @@ TEST(ParamTest, SanitizesParams)
     EXPECT_EQ("a=1;b=2"    , Url::Url("http://foo.com/;a=1;;;;;;b=2;;;").params());
 }
 
+TEST(AbspathTest, BasicPath)
+{
+    EXPECT_EQ("/howdy",
+        Url::Url("http://foo.com/howdy").abspath().path());
+}
+
+TEST(AbspathTest, RepeatedSeparator)
+{
+    EXPECT_EQ("/hello/how/are",
+        Url::Url("http://foo.com/hello//how//are").abspath().path());
+}
+
+TEST(AbspathTest, ParentDirectory)
+{
+    EXPECT_EQ("/how/are",
+        Url::Url("http://foo.com/hello/../how/are").abspath().path());
+}
+
+TEST(AbspathTest, ParentDirectoryWithRepeatedSeparators)
+{
+    EXPECT_EQ("/how/",
+        Url::Url("http://foo.com/hello//..//how/").abspath().path());
+}
+
+TEST(AbspathTest, GrandparentDirectory)
+{
+    EXPECT_EQ("/c",
+        Url::Url("http://foo.com/a/b/../../c").abspath().path());
+}
+
+TEST(AbspathTest, UpMoreLevelsThanSegments)
+{
+    EXPECT_EQ("/c",
+        Url::Url("http://foo.com/../../../c").abspath().path());
```

Comment:
I'm not sure what's the best way to handle this case. It was definitely the case that old web servers once allowed this type of URL to reference paths *outside* of the servers web root, but at some point in time that was (rightly) deemed to be insecure, and thus now as a rule, I'm pretty sure web servers are just going to issue a 404 or 401 for any path that resolves to outside the web root.

In general, these types of URLs are often seen as malicious attempts to bypass security, so I'd be tempted to reject them rather than translate them into something meaningful. The end result may still be sufficient to produce something which might be still be considered malicious. Here's a common example:

`http://example.com/../../../etc/passwd`.

Translating that to:

`http://example.com/etc/passwd` may still get you into hot water for bad bot detection.